[BUG_FIX] Fix resize test

electriclilies opened the pull request:
This PR fixes the bug in #6237. In resize, the method "nearest_neighbor" is not compatible with the coordinate_transform_mode "align_corners", and the method "bilinear" is not compatible with coordinate_transformation_modes other than "align_corners". 

I also added checks to the topi resize method to prevent the user from passing incompatible method and coordinate_transformation_mode combinations into resize.

@mbrookhart @masahi please take a look!

mbrookhart commented on tests/python/relay/test_op_level5.py:
```diff
@@ -67,6 +69,7 @@ def verify_resize(dshape, scale, method, layout):
     for method in ["bilinear", "nearest_neighbor"]:
         for layout in ["NHWC", "NCHW"]:
             verify_resize((1, 4, 4, 4), 2, method, layout)
+            verify_resize((2, 8, 17, 20), 2, method, layout)
```
Can we also add a test case with a different scale?

electriclilies replied:
Just added another scale

mbrookhart commented on python/tvm/topi/image/resize.py:
```diff
@@ -529,8 +529,12 @@ def resize(data, size, layout="NCHW", method="bilinear",
+            (method == "bilinear" and coordinate_transformation_mode != "align_corners")):
+        raise ValueError('Topi Resize does not support the combination of method %s ' \
+                         'and coordinate_transformation_mode %s' %
+                         (method, coordinate_transformation_mode))
```
@masahi any restrictions on bicubic?

masahi replied:
Looking at the test cases in topi https://github.com/apache/incubator-tvm/blob/master/tests/python/topi/python/test_topi_image.py#L69-L83,

I think the valid combination is
* nearest + asymmetric
* bilinear + align_corners or half_pixel

@electriclilies please allow bilinear + half_pixel option.

I don't know about bicubic, since it is not tested. @jwfromm

electriclilies replied:
The bilinear and half_pixel combination is giving me an error in the resize test, so I thought is was another invalid combination. I just tried to replicate the bilinear half_pixel topi test in relay, but I get an error. 

The input to the bilinear and half_pixel combination test_topi_python is verify_resize(4, 16, 16, 16, 32, 32, 'NCHW', "half_pixel", method="bilinear"), which passes in a shape of (4, 16, 16, 16) and resizes it to (4, 16, 32, 32) using the topi resize method. This passes. 

For the relay resize function, if I pass in (4, 16, 16, 16) as the shape, and try to resize it using scale 2, we get an output shape of (4, 16, 32, 32), which is correct, but the test fails the test due output mismatch. I'm not entirely sure what's going on here, I'll look into it more tomorrow and post an update.. if either of you have any insight let me know

masahi replied:
Bilinear + half_pixel should work fine. I guess you are not passing the `half_pixel` option to the reference function here
https://github.com/apache/incubator-tvm/blob/master/tests/python/relay/test_op_level5.py#L54

See https://github.com/apache/incubator-tvm/blob/0a4ee30da727ae1566a0541ba6ed726d48733cc7/python/tvm/topi/testing/bilinear_resize_python.py#L23

masahi commented:
Frontend tests are failing really bad with this change. We need to fix
https://github.com/apache/incubator-tvm/blob/master/python/tvm/relay/frontend/onnx.py#L940

There is also bilinear + align_corners=False combination in tflite test, which translates to bilinear + asymmetric in Relay:
https://github.com/apache/incubator-tvm/blob/77042327d944fbfa243f59740c087d433bc1569e/tests/python/frontend/tflite/test_forward.py#L1058

I don't know what TF does with bilinear + align_corners=False. Also not sure if bilinear + asymmetric in Relay is a valid combination. Maybe @jwfromm @srkreddy1238 have thought on this?

electriclilies commented:
@masahi I changed the onnx frontend to set align_corners to False if the method is 'nearest_neighbor', otherwise set it to true. This got rid of the issue in the onnx importer. 

TFLite is not passing one test, though, because it's trying to use the bilinear + asymmetric combination...

electriclilies commented:
It looks like bilinear + asymmetric passes the resize test, but for some reason it is not tested in the topi test. I added it to the topi test and removed that check from the topi resize op.

masahi commented:
Yes this change now looks good to me. After CI pass we can merge.

electriclilies commented:
@masahi Thanks!

masahi commented:
oh CI is failing because there is nearest + half_pixel combination in pytorch frontend ...

https://github.com/apache/incubator-tvm/blob/939a42b4e976a41e8513b720421d3c3678493715/python/tvm/relay/frontend/pytorch.py#L1542-L1545

@electriclilies can you fix the condition above to 
```
        if method == "nearest_neighbor":
            coord_trans = "asymmetric"
        elif align_corners:
            coord_trans = "align_corners"
        else:
            coord_trans = "half_pixel"
```

Options in resize is a hairy issue, sorry to have you go through this.
